### ISSUE TITLE
ethclient: set independent http.Header for each request.

### ISSUE THREAD

**Issue by storyicon** [ethereum/go-ethereum#26004]
# Rationale

The blockchain gateway we use supports many additional request parameters, which are specified through the headers of the request. For example, we require the client to pass `x-consistent-hash-key` to support consistent hashing. In fact, if there are any additional extensions on the server side, HTTP Headers is the best way at present.

However, I noticed that currently only global headers are supported:
https://github.com/ethereum/go-ethereum/blob/master/rpc/http.go#L215
It cannot set headers for each independent request. (For example, only for `eth_sendRawTransaction`, without affecting other read requests)

# Implementation

I do not deny that there are other ways to achieve the goal, but if there are the following mechanisms, it will be very convenient:
```
client, _ := ethclient.Dial("...")

header := http.Header{}
header.Set("x-consistent-hash", "...")
ctx := ethclient.WithHeaders(context.Background(), header)
if err := client.SendTransaction(ctx, tx); err != nil {
	panic(err)
}
```

If I propose a PR to add this feature, will it be merged?

**Comment by jsvisa** [comment]
Maybe you can check here for an example:


https://github.com/ethereum/go-ethereum/blob/fb75f11e87420ec25ff72f7eeeb741fa8974e87e/rpc/client_opt_test.go#L13-L25

initialize a new instance for sending the transaction

**Comment by storyicon** [comment]
Thank you, but as mentioned in my description:
```
It cannot set headers for each independent request. (For example, only for eth_sendRawTransaction, without affecting other read requests)
```
This method has relatively large limitations.
In contrast, in the Client design of HTTP and GRPC, the entrance to set up Header for each request is provided. I think this is a very basic design problem.

**Comment by lightclient** [comment]
The way I've got around this in the past is by calling the inner `rpc.Client`'s `SetHeader` function. You could easily wrap the `ethclient` in a way that this is done automatically for certain calls. If we were to accept a PR for this, I think the best path would be to follow the design of `rpc.Client` and create a `SetHeader` function.

**Comment by storyicon** [comment]
Thanks @lightclient 
This design still sets the same header for all requests, which has great limitations:
![image](https://user-images.githubusercontent.com/29772821/196880273-b027f9ad-97df-4983-a36c-f9b38349b58c.png)
In a consistent hash scenario, each request may require a different consistent hash. 
There are more scenes. It should be a common requirement.

**Comment by lightclient** [comment]
You can call `SetHeader` again after the call is made to unset the header.

**Comment by storyicon** [comment]
Sorry @lightclient , I don't think this is a robust solution. This is based on the assumption that `SetHeader` and `doRequest` are atomic operations, which will cause confusion when concurrent requests are made.

**Comment by storyicon** [comment]
This issue is now resolved by #26023.
